### DISCUSSION TITLE
DCT to an image

### DISCUSSION THREAD

**Discussion by SheepDomination** [ImageMagick/ImageMagick#6666] [Digital Image Processing]
Hi, I recently read about DCT and I know it has been discussed regarding ImagMagick; although how do I add it to an image ?

**Comment by fmw42** [comment]
Your request is vague.  Please be more specific.  What aspect of DCT do you want?  Post and example. Imagemagick does not have a function to create DCT although I believe it is used in JPG compression. IM can create an FFT.

**Reply by SheepDomination** [reply]
Hi, I didn't know there is more then one aspect in my searching I could only find one type of examples ?

How is FFT different and does it work with anything other then JPEG ?

**Reply by fmw42** [reply]
This question is not very clear.  Please explain your need for DCT or how it would be used.  DCT and FFT are related but used differently.  FFT is not used in JPG.  DCT is used in JPG compression.  How do you need to use DCT?  Please elaborate your need for DCT

**Reply by SheepDomination** [reply]
My need is to water mark and image so that it cannot be modified or altered and the water mark will retain.

**Reply by fmw42** [reply]
How does DCT apply to watermarking.  You do not need compression to watermark.

**Reply by SheepDomination** [reply]
DCT for Image Watermarking, when you search for this; there are many papers on this subject.

**Reply by fmw42** [reply]
According to http://www.ee.sunysb.edu/~cvl/ese558/s2005/Reports/Lin%20Liu/ese558report_LinLiu.pdf, you can use FFT in place of DCT.  Imagemagick has FFT, but not DCT.  Further reading indicates that the image and watermark have to be separated into non-overlapping 8x8 blocks and combined in the DCT domain and then IDCT to bring it back to the spatial domain.  This would be very intensive on large images if one had to script it.  Imagemagick does not have such an algorithm already coded.  You can watermark normally with -compose modulate  -composite command.  See https://imagemagick.org/Usage/compose/#watermark

**Reply by SheepDomination** [reply]
Is this technique -compose modulate placing a water marked image based on the value of the brightness of the image based on the setting value. 
In other words, if I set the brightness and contrast to 34/60 then the water mark will be placed in those areas of the image ?

**Reply by fmw42** [reply]
Placement is controlled by the -geometry +X+Y and has nothing to do with brightness/contrast of the image or watermark.  It is done in the spatial domain, not the DCT nor FFT domain.  It is a simple blend whose amount depends upon the brightness.  The saturation controls whether and how much the overlay (watermark image) shows in color or grayscale.  See the examples in the link that I provided.

**Reply by SheepDomination** [reply]
It appears that based on the one link that one can extract the watermark from the image. I'm concerned over algorithms such as content aware fill in photoshop. If someone takes a part of the image they want to watermark the image out, using this method; it would be nice if the watermark would still appear on the content aware filled area.

**Reply by fmw42** [reply]
Sorry, I do not know anything about content aware fill in photoshop.  I will leave any further discussion to the IM developers. But please make your question as clear as possible and perhaps show an example of what you want or do not want.

**Reply by fmw42** [reply]
OK found it at https://helpx.adobe.com/photoshop/using/content-aware-fill.html#content-aware-fill
 
Sorry, but I do not understand your question.  Are you trying to add a watermark or remove one.  Imagemagick does not have any method of automatically removing it, such as inpainting.

**Reply by SheepDomination** [reply]
I want to only add a watermark. Although the page you found on content aware fill, it can take a part of an image and remove a part and fill it with another part of the same image. 

What I want in hopes that FFT can do such is if someone uses this technology on a JPEG,PNG as to remove a watermark it will always remain assuming multiple watermarks are on the image and if one didn't exist in the spot to which this technology content aware fill was used on that one is always within the image and will reveal it self so to speak ?

**Reply by fmw42** [reply]
Imagemagick does not have such an FFT tool.  Imagemagick can only add a watermark to a region. It cannot do content aware fill.

**Comment by snibgo** [comment]
I haven't experimented with this, but here is a quick demonstration of using FFT for watermarks. Windows BAT syntax.

We embed the word "snibgo" in the frequency domain of `toes.png`.
```
%IM7DEV%magick ^
  toes.png ^
  -alpha off ^
  -fft ^
  ( -clone 0 ^
    -gravity NorthWest ^
    -pointsize 25 ^
    -fill Black ^
    -annotate +10+10 "snibgo" ^
  ) ^
  -swap 0,2 +delete ^
  -ift ^
  -alpha off ^
  -crop 267x233+0+0 +repage ^
  x.png
```
`x.png` is:
![x](https://github.com/ImageMagick/ImageMagick/assets/33812368/a8fdc7bf-98f3-4c2f-9e37-7ba282a8375f)

The watermark is distributed throughout the image. It isn't visible, and can't be removed by inpainting.


We can detect the watermark like this:
```
%IM7DEV%magick ^
  x.png ^
  -fft ^
  -delete 1 ^
  -crop 45x25%%+0+0 ^
  -auto-level -auto-gamma ^
  f0.png
```
![f0](https://github.com/ImageMagick/ImageMagick/assets/33812368/5f0fa21b-250d-4f09-b60b-de48becaf448)

A thief may try simple techniques to defeat the watermark, for example:
```
%IM7DEV%magick ^
  x.png ^
  -crop 200x200+20+20 +repage ^
  -quality 40 ^
  x.jpg
```
![x](https://github.com/ImageMagick/ImageMagick/assets/33812368/73d15420-e6af-4ff3-b347-930ef8abbae1)


Did the thief succeed? Is the watermark still detectable?
```
%IM7DEV%magick ^
  x.jpg ^
  -fft ^
  -delete 1 ^
  -crop 45x25%%+0+0 ^
  -auto-level -auto-gamma ^
  xf0.png
```
`xf0.png` is:
![f0](https://github.com/ImageMagick/ImageMagick/assets/33812368/b10ed60e-61da-4f4a-b893-0414c757257f)

Of course, if a thief knows this method has been used, it can be defeated by `-fft` then overpainting the logo with black, then `-ift`.

**Reply by fmw42** [reply]
Excellent Snibgo.  Thanks. I did not have time to script anything.  Been busy with other things.

One thing that I see in some of the literature is that they do this in blocks and shuffle the blocks. To recover, one needs the key to the shuffling.  (I do not expect you to do that here. Just thought I would pass that on).

**Comment by SheepDomination** [comment]
@snibgo That is well embedded, I don't even see the watermark in the first image. If you embed the image in a specific spot as well as other spots content aware fill won't have any effect because you will be removing part of the image, destroying it. 

How about preventing the image from being defeated.

**Comment by SheepDomination** [comment]
Also how about showing the watermark somewhere on the image but hiding the rest ?

The technique you showed @snibgo is good, What do you mean by overpainting the logo with black ?

**Comment by snibgo** [comment]
I made a mistake above. Suppose the thief tries to remove the watermark by adding noise:
```
%IM7DEV%magick ^
  x.png ^
  -attenuate 0.5 +noise Uniform ^
  f.png
```
We can still detect the watermark like this:
```
%IM7DEV%magick ^
  f.png ^
  -fft ^
  -delete 1 ^
  -crop 45x25%%+0+0 ^
  -auto-level -auto-gamma ^
  xf2.png
```
`xf2.png` shows the watermark:

![xf2](https://github.com/ImageMagick/ImageMagick/assets/33812368/4ed691e8-8adf-4ae4-8292-51e23293f2a6)

This simple scheme can be defeated by overpainting (eg drawing a solid black rectangle, or "content-aware fill") in the frequency domain on the FFT image.

An ideal watermark would not be detectable or removable by a thief, but the image owner could still prove that an image contains the watermark, even after the thief adds noise etc. One way is for the owner to scramble the original image with a private key before applying the watermark, then de-scrambling the result, and publishing that. If the thief doesn't know the key, he cannot detect the watermark, or remove it. A public/private key would be better, so *anyone* with the public key could prove the modified image contained the logo, and could prove the owner of the private key must have created it.

The watermark is embedded in the *frequency* domain, so is not visible in the image. Another watermark could be added in the spatial domain, so it would be visible.

I haven't investigated, researched or coded this.

**Reply by SheepDomination** [reply]
> One way is for the owner to scramble the original image with a private key before applying the watermark, then de-scrambling the result, and publishing that. If the thief doesn't know the key, he cannot detect the watermark, or remove it. A public/private key would be better, so anyone with the public key could prove the modified image contained the logo, and could prove the owner of the private key must have created it.

How could this be accomplished, do you know what this may be called 

Can one place the watermark in a specific area of the image using frequency domain or FFT ?

Spatial domain watermark would simple help in knowing if the watermark is placed, whether that is in the middle of the image or somewhere on the image; but from my understanding it's not in frequency domain aka within layers of the image ?

**Reply by fmw42** [reply]
I believe it can be placed anywhere in the FFT image.

Scrambling can be done via the Arnold Algorithm.  See https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwjntvWnssSBAxVZCTQIHSjzBu8QFnoECBQQAQ&url=https%3A%2F%2Fwww.atlantis-press.com%2Farticle%2F10512.pdf&usg=AOvVaw23IVxtoNVG_8HDJVqF046e&opi=89978449 and https://www.alliance.edu.in/aicaam-conference-proceedings/Papers/Image-Scrambling-through-Two-Level-Arnold-Transform.pdf and https://en.wikipedia.org/wiki/Arnold%27s_cat_map

**Reply by SheepDomination** [reply]
ImageMagick does not have the Arnold Algorithm ?

**Comment by snibgo** [comment]
"Spatial domain" means the pixel at a particular location represents the colour at the same location.

"Frequency domain" means the pixel at a particular location represents the magnitude and phase of a particular frequency. For more details, see https://imagemagick.org/Usage/fourier/

An image can be represented with either method. We normally use the spatial domain. Any changes we make to pixels will show as changes to that part of the image.

When we work in the frequency domain, any changes (to the magnitude or phase) may show as changes in any (spatial) part of the image, or in all parts.

I made the changes near a corner of the frequency magnitude. Near the corners, which represent high frequencies, the magnitude is close to zero. I made some pixels, in the shape of "snibgo", exactly black. The changes in the spatial domain are not visible. If I made that change near the centre, which represents low frequencies and magnitudes are not near zero, the changes to the spatial image would be obvious.

**Reply by SheepDomination** [reply]
How do you know what part of the image is low or high frequency; as so that I know if I want to place the watermark in specific areas ?

These methods, especially FFT work on JPG,PNG; and GIF. Therefore what about TIF,EXR,TGA; same or is it different especially if the image is 32 bit ?

**Comment by snibgo** [comment]
Each iteration of Arnold's Cat Map could be done with `-fx`.

However, I don't think Arnold's Cat Map is a good choice for scrambling in this context. The Wikipedia article says, "The number of iterations needed to restore the image can be shown never to exceed 3N" for an image of NxN pixels. So there are only 3N possible key values, which is far too few for cryptographic security.

A better choice might be XOR the image with some other random image. That other image is then the private key. Each (colour) pixel of that key has at least 2^24 possible values. 

> How do you know what part of the image is low or high frequency; as so that I know if I want to place the watermark in specific areas ?

I don't understand the question. With this scheme, the watermark is not placed in a specific area of the spatial image.

> These methods, especially FFT work on JPG,PNG; and GIF. Therefore what about TIF,EXR,TGA; same or is it different especially if the image is 32 bit ?

FFT works on an image. It doesn't matter what file format the image came from.

**Comment by SheepDomination** [comment]
Is it possible to place a watermark in a specific area of the image or it does this automatically based on the frequency of the image ?

XOR ?

**Comment by fmw42** [comment]
Using the FFT method, the image is placed in the FFT image and as such is distributed all over the input image. So you cannot place the watermark in a specific place in the input image. It is placed everywhere in the input image in an invisible and non-readable form.

**Reply by fmw42** [reply]
The Arnold method is used in watermarking if one shuffles blocks of pixels. Using blocks rather than individual pixels makes it less time consuming the larger the blocks are. One decides on a compromise size of blocks that is not too larger and not too small such that the time it takes to shuffle and reconstitute is not too time consuming.

**Reply by SheepDomination** [reply]
ImageMagick cannot do XOR or Arnold, although XOR is better choice ?

**Comment by snibgo** [comment]
One iteration of Arnold's Cat Map, to shuffle individual pixels, can be done like this:
```
magick in.png   -virtual-pixel Tile -fx "p{2*i+j,i+j}" out.png
```
For multiple iterations, either repeat the `-fx` or put it in a shell loop.

XOR can be done like this:
```
magick in1.png in2.png -fx "(((QuantumRange*u)&(QuantumRange*(1-v)))|((QuantumRange*(1-u))&(QuantumRange*v)))/QuantumRange" out.png
```

**Reply by SheepDomination** [reply]
How would you show the watermark on the image once it's embedded so you can verify if it's yours ?

Therefore with XOR you need two images as the watermark for secure watermarking that cannot be hacked in any way ?

**Comment by fmw42** [comment]
One can also use Imagemgick -stegano to distribute a message throughout an image.  See https://imagemagick.org/Usage/transform/#stegano
